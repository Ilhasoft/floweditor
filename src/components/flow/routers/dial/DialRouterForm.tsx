import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import Dialog, { ButtonSet, Tab } from 'components/dialog/Dialog';
import { hasErrors, renderIssues } from 'components/flow/actions/helpers';
import { RouterFormProps } from 'components/flow/props';
import { nodeToState, stateToNode } from 'components/flow/routers/dial/helpers';
import { createResultNameInput } from 'components/flow/routers/widgets';
import TextInputElement from 'components/form/textinput/TextInputElement';
import TypeList from 'components/nodeeditor/TypeList';
import { FormState, StringEntry } from 'store/nodeEditor';
import { Alphanumeric, Required, StartIsNonNumeric, validate } from 'store/validators';
import i18n from 'config/i18n';
import CheckboxElement from 'components/form/checkbox/CheckboxElement';

export interface DialRouterFormState extends FormState {
  phone: StringEntry;
  resultName: StringEntry;
}

export default class DialRouterForm extends React.Component<RouterFormProps, DialRouterFormState> {
  constructor(props: RouterFormProps) {
    super(props);

    this.state = nodeToState(this.props.nodeSettings);

    bindCallbacks(this, {
      include: [/^on/, /^handle/]
    });
  }

  private handlePhoneUpdated(value: string): void {
    const phone = validate(i18n.t('forms.phone_number', 'Phone Number'), value, [Required]);
    this.setState({
      phone: phone,
      valid: this.state.valid && !hasErrors(phone)
    });
  }

  private handleUpdateResultName(value: string): void {
    const resultName = validate(i18n.t('forms.result_name', 'Result Name'), value, [
      Alphanumeric,
      StartIsNonNumeric
    ]);
    this.setState({
      resultName,
      valid: this.state.valid && !hasErrors(resultName)
    });
  }

  private handleSave(): void {
    if (this.state.valid) {
      this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
      this.props.onClose(false);
    }
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.ok', 'Ok'), onClick: this.handleSave },
      secondary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true)
      }
    };
  }

  public renderEdit(): JSX.Element {
    const typeConfig = this.props.typeConfig;

    const advanced: Tab = {
      name: i18n.t('forms.advanced', 'Advanced'),
      body: (
        <CheckboxElement
          name={i18n.t('forms.all_destinations', 'All Destinations')}
          title={i18n.t('forms.all_destinations', 'All Destinations')}
          // todo labelClassName={styles.checkbox}
          checked={false} // todo checked={this.state.sendAll}
          description={i18n.t(
            'forms.all_destinations_description',
            "Send a message to all destinations known for this contact. If you aren't sure what this means, leave it unchecked."
          )}
          // todo onChange={this.handleSendAllUpdate}
        />
      )
      // todo checked: this.state.sendAll
    };

    const tabs = [advanced];

    return (
      <Dialog
        title={typeConfig.name}
        headerClass={typeConfig.type}
        buttons={this.getButtons()}
        tabs={tabs}
      >
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        <p>Enter the phone number to dial</p>
        <TextInputElement
          name="phone"
          placeholder={i18n.t('forms.phone_number', 'Phone Number')}
          showLabel={false}
          autocomplete={true}
          onChange={this.handlePhoneUpdated}
          entry={this.state.phone}
        />
        {createResultNameInput(this.state.resultName, this.handleUpdateResultName)}
        {renderIssues(this.props)}
      </Dialog>
    );
  }

  public render(): JSX.Element {
    return this.renderEdit();
  }
}
