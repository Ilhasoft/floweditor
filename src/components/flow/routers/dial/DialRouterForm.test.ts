import { RouterFormProps } from 'components/flow/props';
import DialRouterForm from 'components/flow/routers/dial/DialRouterForm';
import { Types } from 'config/interfaces';
import { composeComponentTestUtils, mock } from 'testUtils';
import { getRouterFormProps, createMatchRouter } from 'testUtils/assetCreators';
import * as utils from 'utils';
import { createUUID } from 'utils';

mock(utils, 'createUUID', utils.seededUUIDs());

const routerNode = createMatchRouter([]);
routerNode.ui = {
  position: { left: 0, top: 0 },
  type: Types.wait_for_dial
};

const { setup } = composeComponentTestUtils<RouterFormProps>(
  DialRouterForm,
  getRouterFormProps(routerNode)
);

describe(DialRouterForm.name, () => {
  it('should render', () => {
    const { wrapper } = setup(true);
    expect(wrapper).toMatchSnapshot();
  });

  describe('updates', () => {
    it('should save changes', () => {
      const { instance, props } = setup(true, {
        $merge: { updateRouter: jest.fn(), onClose: jest.fn() }
      });

      instance.handleUpdateResultName('Dial Result');
      instance.handlePhoneUpdated('@fields.supervisor');

      expect(instance.state).toMatchSnapshot();

      instance.handleSave();
      expect(props.onClose).toHaveBeenCalled();
      expect(props.updateRouter).toHaveBeenCalled();
      expect(props.updateRouter).toMatchCallSnapshot();
    });

    it('should cancel', () => {
      const { instance, props } = setup(true, {
        $merge: { updateRouter: jest.fn(), onClose: jest.fn() }
      });

      instance.handlePhoneUpdated('+1234567890');
      instance.getButtons().secondary.onClick();
      expect(props.onClose).toHaveBeenCalled();
      expect(props.updateRouter).not.toHaveBeenCalled();
    });
  });
});
