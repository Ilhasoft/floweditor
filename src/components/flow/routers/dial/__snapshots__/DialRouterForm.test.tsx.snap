// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialRouterForm should render 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="header"
      >
        <div
          class="title_container"
        >
          <div
            class="title u font secondary title-sm black"
          >
            Wait for Forwarded Call
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <div
            class="color-neutral-cloudy"
          >
            When a contact arrives at this point in your flow...
          </div>
          <div>
            <div
              data-use-vue-component-wrap=""
              style="all: unset;"
            >
              <div
                class="unnnic-autocomplete-select bold_placeholder"
                data-passed-props="[object Object]"
                data-testid="temba_select_type"
                data-v-4e186498=""
              >
                <div
                  class="unnnic-form sm"
                  data-v-48c0c241=""
                  data-v-4e186498=""
                >
                  <!---->
                  <div
                    class="unnnic-form-input"
                    data-v-2d3b8b50=""
                    data-v-48c0c241=""
                    mask=""
                    style="position: relative;"
                    v-bind="[object Object]"
                  >
                    <input
                      class="input size-sm normal has-icon-left"
                      data-v-2d3b8b50=""
                      data-v-f03c53b0=""
                      iconleft="search-1"
                      placeholder="Wait for forwarded call to someone else"
                      size="0"
                      type="text"
                      v-bind="[object Object]"
                    />
                    <span
                      class="icon-left unnnic-icon unnnic-icon__size--xs unnnic-icon-scheme--neutral-cloudy"
                      data-v-2d3b8b50=""
                      data-v-a1936218=""
                    >
                      <svg
                        fill="none"
                        viewBox="0 0 40 40"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          class="primary"
                          d="M34.0615 35.0005C33.8116 35.0005 33.5754 34.903 33.3992 34.7256L25.2575 26.5826C24.3065 27.3961 23.2355 28.0597 22.067 28.5559C20.5312 29.2082 18.9178 29.5394 17.272 29.5394C15.7011 29.5394 14.1578 29.2382 12.6819 28.6421C9.64389 27.4161 7.26574 25.0817 5.98481 22.0662C4.70388 19.052 4.67263 15.7178 5.89858 12.6811C7.12452 9.64433 9.45894 7.26492 12.4732 5.98399C14.009 5.33041 15.6211 5.00049 17.2682 5.00049C18.8391 5.00049 20.3825 5.30291 21.8583 5.89776C24.8963 7.12371 27.2745 9.45812 28.5554 12.4736C29.8363 15.4891 29.8676 18.822 28.6416 21.8588C28.1393 23.1035 27.4469 24.2444 26.5809 25.2554L34.7251 33.4009C35.09 33.7658 35.09 34.3606 34.7251 34.7268C34.5477 34.903 34.3115 35.0005 34.0615 35.0005ZM17.2682 6.87502C15.8748 6.87502 14.5077 7.15495 13.2068 7.70856C10.6524 8.79454 8.67413 10.809 7.63564 13.3821C6.59715 15.9552 6.6234 18.7783 7.70813 21.3327C8.7941 23.887 10.8086 25.8653 13.3817 26.9038C14.6326 27.4086 15.9411 27.6636 17.2707 27.6636C18.6641 27.6636 20.0313 27.3836 21.3322 26.83C22.5494 26.3127 23.6404 25.5828 24.5739 24.6606C24.5864 24.6431 24.6002 24.6281 24.6139 24.6143C24.6339 24.5943 24.6526 24.5781 24.6701 24.5656C25.6236 23.6021 26.381 22.4499 26.9021 21.1565C27.9406 18.5833 27.9143 15.7603 26.8283 13.2059C25.7436 10.6516 23.7279 8.67332 21.1548 7.63483C19.9063 7.13121 18.5979 6.87502 17.2682 6.87502Z"
                          fill="#3B414D"
                        />
                      </svg>
                    </span>
                    <!---->
                  </div>
                  <!---->
                </div>
                <!---->
              </div>
            </div>
          </div>
        </div>
        <p>
          Enter the phone number to dial
        </p>
        <div
          data-use-vue-component-wrap=""
          style="all: unset;"
        >
          <div
            class="unnnic-autocomplete-select completionInput"
            data-passed-props="[object Object]"
            data-testid="phone"
            data-v-4e186498=""
          >
            <div
              class="unnnic-form sm"
              data-v-48c0c241=""
              data-v-4e186498=""
            >
              <!---->
              <div
                class="unnnic-form-input"
                data-v-2d3b8b50=""
                data-v-48c0c241=""
                mask=""
                style="position: relative;"
                v-bind="[object Object]"
              >
                <input
                  class="input size-sm normal"
                  data-v-2d3b8b50=""
                  data-v-f03c53b0=""
                  placeholder="Phone number"
                  size="0"
                  type="text"
                  v-bind="[object Object]"
                />
                <!---->
                <!---->
              </div>
              <!---->
            </div>
            <!---->
          </div>
        </div>
        <div
          class="optional_text_input "
        >
          <div
            data-use-vue-component-wrap=""
            style="all: unset;"
            unnnic-input="true"
          >
            <div
              class="unnnic-input size-sm"
              data-passed-props="[object Object]"
              data-testid="Save as result"
              data-v-391ed932=""
            >
              <span
                class="label unnnic-font secondary color-neutral-cloudy body-md"
                data-v-391ed932=""
              >
                 Save as result 
              </span>
              <div
                class="input-container"
                data-v-391ed932=""
              >
                <!---->
                <input
                  class="unnnic-font secondary"
                  data-v-391ed932=""
                  maxlength="64"
                  placeholder="Result"
                  type="text"
                />
                <!---->
                <!---->
              </div>
              <!---->
            </div>
          </div>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="buttons_wrapper"
          >
            <div
              class="right_buttons"
            >
              <div
                data-use-vue-component-wrap=""
              >
                <button
                  class="unnnic-button unnnic-button--size-large unnnic-button--terciary unnnic-button-scheme--"
                  data-passed-props="[object Object]"
                  data-v-00f478da=""
                  style="margin-left: 0px; margin-top: 0px;"
                >
                  <!---->
                  <!---->
                  <!---->
                  <span
                    class="unnnic-button__label"
                    data-v-00f478da=""
                  >
                     Cancel 
                  </span>
                  <!---->
                </button>
              </div>
              <div
                data-use-vue-component-wrap=""
              >
                <button
                  class="unnnic-button unnnic-button--size-large unnnic-button--secondary unnnic-button-scheme--"
                  data-passed-props="[object Object]"
                  data-v-00f478da=""
                  style="margin-top: 0px;"
                >
                  <!---->
                  <!---->
                  <!---->
                  <span
                    class="unnnic-button__label"
                    data-v-00f478da=""
                  >
                     Ok 
                  </span>
                  <!---->
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`DialRouterForm should update wait when phone number is changed 1`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
        },
        Object {
          "uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
        },
        Object {
          "uuid": "6c22884b-0e35-4206-982e-18320691eda9",
        },
        Object {
          "uuid": "0c8c9239-9b59-473a-a229-d8a9693be270",
        },
        Object {
          "uuid": "6b59e863-2e14-442b-9987-aaa284871b8e",
        },
        Object {
          "uuid": "c5722340-f7e1-4d99-ac5e-3a6e59845111",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "answered",
            ],
            "category_uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
            "type": "has_only_text",
            "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
          },
          Object {
            "arguments": Array [
              "no answer",
            ],
            "category_uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
            "type": "has_only_text",
            "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
          },
          Object {
            "arguments": Array [
              "busy",
            ],
            "category_uuid": "a62e8284-49f8-4292-adbc-d5f6607f85c5",
            "type": "has_only_text",
            "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
          },
          Object {
            "arguments": Array [
              "failed",
            ],
            "category_uuid": "9adce29f-da95-4182-9cb3-0f1e3bf714e5",
            "type": "has_only_text",
            "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
            "name": "Answered",
            "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
          },
          Object {
            "exit_uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
            "name": "No Answer",
            "uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
          },
          Object {
            "exit_uuid": "6c22884b-0e35-4206-982e-18320691eda9",
            "name": "Busy",
            "uuid": "a62e8284-49f8-4292-adbc-d5f6607f85c5",
          },
          Object {
            "exit_uuid": "0c8c9239-9b59-473a-a229-d8a9693be270",
            "name": "Failed",
            "uuid": "9adce29f-da95-4182-9cb3-0f1e3bf714e5",
          },
          Object {
            "exit_uuid": "6b59e863-2e14-442b-9987-aaa284871b8e",
            "name": "Other",
            "uuid": "00dd2423-0a34-4ee2-b97d-839fb2d593ef",
          },
          Object {
            "exit_uuid": "c5722340-f7e1-4d99-ac5e-3a6e59845111",
            "name": "No Response",
            "uuid": "f05fa107-03db-4315-a211-53da63603e6f",
          },
        ],
        "default_category_uuid": "f05fa107-03db-4315-a211-53da63603e6f",
        "operand": "@(default(resume.dial.status, \\"\\"))",
        "result_name": "Dial Result",
        "type": "switch",
        "wait": Object {
          "phone": "0979123456",
          "type": "dial",
        },
      },
      "uuid": "141e0356-9fdd-47d3-a12d-16efc5b45276",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "wait_for_dial",
    },
  },
]
`;
